modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/TraceableIterator.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.tracing;
+
+import java.util.Iterator;
+import org.apache.ignite.internal.processors.tracing.MTC.TraceSurroundings;
+
+import static org.apache.ignite.internal.processors.tracing.SpanTags.ERROR;
+
+/**
+ * Represents wrapper which gives an ability to trace usage of the iterator.
+ */
+public class TraceableIterator<T> implements Iterator<T> {
+    /** Iterator to which all calls will be delegated. */
+    private final Iterator<T> iter;
+
+    /** Span in which context iterator should run. */
+    private final Span span;
+
+    /**
+     * @param iter Iterator to which all calls will be delegated.
+     * @param span Span in which context iterator should run.
+     */
+    public TraceableIterator(Iterator<T> iter, Span span) {
+        this.iter = iter;
+        this.span = span;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() {
+        try (TraceSurroundings ignored = MTC.supportContinual(span)) {
+            return iter.hasNext();
+        }
+        catch (Throwable e) {
+            span.addTag(ERROR, e::getMessage);
+
+            throw e;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public T next() {
+        try (TraceSurroundings ignored = MTC.supportContinual(span)) {
```

Comment:
What's the point of re-supporting span from the constructor on every iteration? Seems that adding log-points will be even more useful.
Here's the test that I've used:
```
    @Test
    public void testFooBar() throws Exception {
        try (MTC.TraceSurroundings ignored = MTC.support(ignite(0).context().tracing().create(SpanType.TX))) {

            Set<String> set = Stream.of("a", "b").collect(Collectors.toSet());

            Iterator iter = new TraceableIterator<>(set.iterator(), MTC.span());

            while (iter.hasNext())
                iter.next();
        }
        
        handler().flush();
    }
```
As a result I got one simple span without additional meta, exact the same as the one without iterations. 
However in case of using:
```
    /** {@inheritDoc} */
    @Override public boolean hasNext() {
        try {
            MTC.span().addLog(()-> "hasNext()");

            return iter.hasNext();
        }
        catch (Throwable e) {
            span.addTag(ERROR, e::getMessage);

            throw e;
        }
    }

    /** {@inheritDoc} */
    @Override public T next() {
        try {
            MTC.span().addLog(()-> "next()");

            return iter.next();
        }
        catch (Throwable e) {
            span.addTag(ERROR, e::getMessage);

            throw e;
        }
    }
```
we will populate original parent span with log-points meta.

In general I believe that in some cases iteration might be interesting and important process that deserves specific new span for every iteration step, but in most cases log points are fine here. What I mean here is that tracing iterations is an in-dependable process of any upwards semantic, it could be used both within SQL query processing or Transactions or any other stuff. 

All in all I believe that we might introduce new scope for iteration that might be enabled with included-scopes logic and populate whatever spans with log points.

```
    /** {@inheritDoc} */
    @Override public T next() {
        try (TraceSurroundings ignored = MTC.support(ctx.tracing().create(ITER_NEXT, MTC.span()))) {
            MTC.span().addLog(()-> "next()");

            return iter.next();
        }
        catch (Throwable e) {
            span.addTag(ERROR, e::getMessage);

            throw e;
        }
    }
```

Reply:
The main reason to re-supporting span is that call of the "next" method from the user thread may cause the next page to be fetched from the remote node. And it should be traced in context of the initial query. If we will only use logging, than tracing of page fetch will be lost.

Reply:
>And it should be traced in context of the initial query. If we will only use logging, than tracing of page fetch will be lost.
Why? (We might use previously stored span instead MTC.span() of course)

Reply:
Let me rephrase. The user calls the `next ` method from received cursor in his thread. If during its execution some other spans will be generated(e.g. span for next page request processing on a remote node), it will be detached from the initial query span. Logging gives us an ability to track the fact that the next method was called, but I'd like to show what is going during method execution. Something like this.
![image](https://user-images.githubusercontent.com/32207922/86919010-33362980-c130-11ea-8abd-b9a26626f284.png)